File: core/src/main/java/in/testpress/ui/BaseToolBarActivity.java
package in.testpress.ui;

import android.content.Intent;
import android.os.Bundle;
import androidx.annotation.StringRes;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import android.view.MenuItem;
import android.widget.ImageView;

import com.nostra13.universalimageloader.core.DisplayImageOptions;
import com.nostra13.universalimageloader.core.ImageLoader;

import in.testpress.R;
import in.testpress.core.TestpressSdk;
import in.testpress.core.TestpressSession;
import in.testpress.network.RetrofitCall;
import in.testpress.util.CommonUtils;
import in.testpress.util.ImageUtils;

import static android.view.WindowManager.LayoutParams.FLAG_SECURE;
import static in.testpress.core.TestpressSdk.ACTION_PRESSED_HOME;

/**
 * Base activity used to support the toolbar & handle backpress.
 * Activity that extends this activity must needs to include the #layout/testpress_toolbar
 * in its view.
 */
public abstract class BaseToolBarActivity extends AppCompatActivity {

    public static final String ACTIONBAR_TITLE = "title";
    protected ImageView logo;
    private TestpressSession session;

    @Override
    public void setContentView(final int layoutResId) {
        session = TestpressSdk.getTestpressSession(this);
        if (session != null && session.getInstituteSettings().isScreenshotDisabled()) {
            getWindow().setFlags(FLAG_SECURE, FLAG_SECURE);
        }
        super.setContentView(layoutResId);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        logo = findViewById(R.id.toolbar_logo);
        setSupportActionBar(toolbar);
        //noinspection ConstantConditions
        getSupportActionBar().setHomeButtonEnabled(true);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
    }

    public void setActionBarTitle(@StringRes int titleId) {
        //noinspection ConstantConditions
        getSupportActionBar().setTitle(titleId);
    }

    public void setActionBarTitle(String title) {
        //noinspection ConstantConditions
        getSupportActionBar().setTitle(title);
    }

    public void setLogo() {
        getSupportActionBar().setTitle("");
        if (session == null || session.getInstituteSettings() == null) {
            return;
        }
        String url = session.getInstituteSettings().getAppToolbarLogo();
        ImageLoader imageLoader = ImageUtils.initImageLoader(this);
        DisplayImageOptions options = new DisplayImageOptions.Builder()
                .cacheInMemory(true)
                .cacheOnDisk(true)
                .build();
        imageLoader.displayImage(url, logo, options);
    }

    @SuppressWarnings("DanglingJavadoc")
    @Override
    public boolean onOptionsItemSelected(final MenuItem item) {
        if(item.getItemId() == android.R.id.home) {
            /**
             * Set result with home button pressed flag if activity is started by
             * {@link #startActivityForResult}
             *
             * Note: {@link #getCallingActivity} return null if activity is started using
             * {@link #startActivity} instead of {@link #startActivityForResult}
             */
            if (getCallingActivity() == null) {
                onBackPressed();
            } else {
                setResult(RESULT_CANCELED, new Intent().putExtras(getDataToSetResult()));
                super.onBackPressed();
            }
            return true;
        }
        return false;
    }

    @SuppressWarnings("DanglingJavadoc")
    @Override
    public void onBackPressed() {
        /**
         * Set result with home button pressed flag false if activity is started by
         * {@link #startActivityForResult}
         */
        if (getCallingActivity() != null) {
            setResult(RESULT_CANCELED, new Intent().putExtra(ACTION_PRESSED_HOME, false));
        }
        try {
            super.onBackPressed();
        } catch (IllegalStateException e) {
            supportFinishAfterTransition();
        }
    }

    protected Bundle getDataToSetResult() {
        Bundle bundle = new Bundle();
        bundle.putBoolean(ACTION_PRESSED_HOME, true);
        return bundle;
    }

    public RetrofitCall[] getRetrofitCalls() {
        return new RetrofitCall[] {};
    }

    @Override
    protected void onDestroy() {
        CommonUtils.cancelAPIRequests(getRetrofitCalls());
        super.onDestroy();
    }

    @Override
    public void onResume() {
        super.onResume();
        TestpressSession session = TestpressSdk.getTestpressSession(this);
        if (session != null && session.getInstituteSettings().isScreenshotDisabled()) {
            getWindow().setFlags(FLAG_SECURE, FLAG_SECURE);
        }
    }

}

File: core/src/main/java/in/testpress/ui/BaseToolBarActivity.kt
package `in`.testpress.ui

import `in`.testpress.R
import `in`.testpress.core.TestpressSdk
import `in`.testpress.core.TestpressSession
import `in`.testpress.network.RetrofitCall
import `in`.testpress.util.CommonUtils
import `in`.testpress.util.UIUtils
import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.view.WindowManager
import android.widget.ImageView
import androidx.annotation.StringRes
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import `in`.testpress.util.applySystemBarColors

/**
 * Base activity used to support the toolbar & handle backpress.
 * Activity that extends this activity must needs to include the #layout/testpress_toolbar
 * in its view.
 */

open class BaseToolBarActivity: AppCompatActivity() {

    companion object {
        const val ACTIONBAR_TITLE = "title"
    }

    private var session: TestpressSession? = null
    protected lateinit var logo: ImageView
    protected lateinit var toolbar: Toolbar

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        applySystemBarColors(window.decorView.rootView)
    }

    override fun setContentView(layoutResId: Int) {
        super.setContentView(layoutResId)
        preventScreenshot()
        setupActionBar()
    }

    override fun setContentView(view: View?) {
        super.setContentView(view)
        preventScreenshot()
        setupActionBar()
    }

    private fun preventScreenshot() {
        session = TestpressSdk.getTestpressSession(this)
        if (session != null && session!!.instituteSettings.isScreenshotDisabled) {
            window.setFlags(
                WindowManager.LayoutParams.FLAG_SECURE,
                WindowManager.LayoutParams.FLAG_SECURE
            )
        }
    }

    private fun setupActionBar() {
        toolbar = findViewById(R.id.toolbar)
        logo = findViewById(R.id.toolbar_logo)
        setSupportActionBar(toolbar)
        supportActionBar!!.setHomeButtonEnabled(true)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
    }

    open fun setActionBarTitle(@StringRes titleId: Int) {
        supportActionBar!!.setTitle(titleId)
    }

    open fun setActionBarTitle(title: String) {
        supportActionBar!!.title = title
    }

    open fun showLogoInToolbar() {
        supportActionBar!!.title = ""
        if (session == null || session!!.instituteSettings == null) {
            return
        }
        UIUtils.loadLogoInView(logo, this)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            /**
             * Set result with home button pressed flag if activity is started by
             * [.startActivityForResult]
             *
             * Note: [.getCallingActivity] return null if activity is started using
             * [.startActivity] instead of [.startActivityForResult]
             */
            if (callingActivity == null) {
                onBackPressed()
            } else {
                setResult(RESULT_CANCELED, Intent().putExtras(getDataToSetResult()))
                super.onBackPressed()
            }
            return true
        }
        return false
    }

    protected open fun getDataToSetResult(): Bundle {
        return Bundle().apply { putBoolean(TestpressSdk.ACTION_PRESSED_HOME, true) }
    }

    open fun getRetrofitCalls(): Array<RetrofitCall<*>> {
        return arrayOf()
    }

    override fun onBackPressed() {
        /**
         * Set result with home button pressed flag false if activity is started by
         * [.startActivityForResult]
         */
        if (callingActivity != null) {
            setResult(RESULT_CANCELED, Intent().putExtra(TestpressSdk.ACTION_PRESSED_HOME, false))
        }
        try {
            super.onBackPressed()
        } catch (e: IllegalStateException) {
            supportFinishAfterTransition()
        }
    }

    override fun onResume() {
        super.onResume()
        preventScreenshot()
    }

    public override fun onDestroy() {
        CommonUtils.cancelAPIRequests(getRetrofitCalls())
        super.onDestroy()
    }

}